### PR TITLE
[SPARK-13925] [ML] [SparkR] Expose R-like summary statistics in SparkR::glm for more family and link functions

### PR DESCRIPTION
**Pull request by yanboliang**
## What changes were proposed in this pull request?

Expose R-like summary statistics in SparkR::glm for more family and link functions.
Note: Not all values in R [summary.glm](http://stat.ethz.ch/R-manual/R-patched/library/stats/html/summary.glm.html) are exposed, we only provide the most commonly used statistics in this PR. More statistics can be added in the followup work.
## How was this patch tested?

Unit tests.

SparkR Output:

```
Deviance Residuals:
(Note: These are approximate quantiles with relative error <= 0.01)
     Min        1Q    Median        3Q       Max
-0.95096  -0.16585  -0.00232   0.17410   0.72918

Coefficients:
                    Estimate  Std. Error  t value  Pr(>|t|)
(Intercept)         1.6765    0.23536     7.1231   4.4561e-11
Sepal_Length        0.34988   0.046301    7.5566   4.1873e-12
Species_versicolor  -0.98339  0.072075    -13.644  0
Species_virginica   -1.0075   0.093306    -10.798  0

(Dispersion parameter for gaussian family taken to be 0.08351462)

    Null deviance: 28.307  on 149  degrees of freedom
Residual deviance: 12.193  on 146  degrees of freedom
AIC: 59.22

Number of Fisher Scoring iterations: 1
```

R output:

```
Deviance Residuals: 
     Min        1Q    Median        3Q       Max  
-0.95096  -0.16522   0.00171   0.18416   0.72918  

Coefficients:
                  Estimate Std. Error t value Pr(>|t|)    
(Intercept)        1.67650    0.23536   7.123 4.46e-11 ***
Sepal.Length       0.34988    0.04630   7.557 4.19e-12 ***
Speciesversicolor -0.98339    0.07207 -13.644  < 2e-16 ***
Speciesvirginica  -1.00751    0.09331 -10.798  < 2e-16 ***

---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for gaussian family taken to be 0.08351462)

    Null deviance: 28.307  on 149  degrees of freedom
Residual deviance: 12.193  on 146  degrees of freedom
AIC: 59.217

Number of Fisher Scoring iterations: 2
```

cc @mengxr

### CODE REVIEW

File: R/pkg/R/mllib.R [hunk 1]
```diff
@@ -101,12 +101,54 @@ setMethod("summary", signature(object = "GeneralizedLinearRegressionModel"),
...
           })
 
+#' Print the summary of GeneralizedLinearRegressionModel
+#'
+#' @rdname print
+#' @name print.summary.GeneralizedLinearRegressionModel
+#' @export
+print.summary.GeneralizedLinearRegressionModel <- function(x, ...) {
+  x$deviance.resid <- setNames(unlist(approxQuantile(x$deviance.resid, "devianceResiduals",
+    c(0.0, 0.25, 0.5, 0.75, 1.0), 0.0)), c("Min", "1Q", "Median", "3Q", "Max"))
```
Comment (yanboliang) [inline R/pkg/R/mllib.R]: Here we set ```relativeError``` of ```approxQuantile``` to 0.0 which may be very expensive for computing. Should we change to more loose value and document the difference between SparkR and R?

File: R/pkg/R/mllib.R [hunk 2]
```diff
@@ -101,12 +101,54 @@ setMethod("summary", signature(object = "GeneralizedLinearRegressionModel"),
...
+            class(ans) <- "summary.GeneralizedLinearRegressionModel"
+            return(ans)
           })
 
+#' Print the summary of GeneralizedLinearRegressionModel
+#'
+#' @rdname print
+#' @name print.summary.GeneralizedLinearRegressionModel
+#' @export
+print.summary.GeneralizedLinearRegressionModel <- function(x, ...) {
```
Comment (yanboliang) [inline R/pkg/R/mllib.R]: Add this S3 function for formatted output of summary(GeneralizedLinearRegressionModel).

File: R/pkg/R/mllib.R (hunk 1 above)
Comment (mengxr) [inline R/pkg/R/mllib.R]: This is quite expensive. If we just need to show the quartiles, a relative error of `0.01` should work well. I'm not sure whether the `min` and `max` would be correct or not, worthing testing. In the output, we can mention `relative error <= 0.01`.

File: mllib/src/main/scala/org/apache/spark/ml/r/GeneralizedLinearRegressionWrapper.scala [hunk 3]
```diff
@@ -30,19 +30,59 @@ private[r] class GeneralizedLinearRegressionWrapper private (
   private val glm: GeneralizedLinearRegressionModel =
     pipeline.stages(1).asInstanceOf[GeneralizedLinearRegressionModel]
 
+  lazy val rFeatures: Array[String] = if (glm.getFitIntercept) {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/r/GeneralizedLinearRegressionWrapper.scala]: We can remove the `r` prefix and just call it `features`. The whole class is a R wrapper.

File: R/pkg/inst/tests/testthat/test_mllib.R [hunk 4]
```diff
@@ -77,6 +77,55 @@ test_that("glm and predict", {
...
+  expect_equal(stats$null.deviance, rStats$null.deviance)
+  expect_equal(stats$deviance, rStats$deviance)
+  expect_equal(stats$df.null, rStats$df.null)
+  expect_equal(stats$df.residual, rStats$df.residual)
+  expect_equal(stats$aic, rStats$aic)
+
+  # Test summary works on base GLM models
+  baseModel <- stats::glm(Sepal.Width ~ Sepal.Length + Species, data = iris)
+  baseSummary <- summary(baseModel)
+  expect_true(abs(baseSummary$deviance - 12.19313) < 1e-4)
```
Comment (mengxr) [inline R/pkg/inst/tests/testthat/test_mllib.R]: `12.19313` -> `baseSummary$deviance`

**Comment by mengxr** [comment]
LGTM except `relativeError` and a magic number in test. This PR looks great!!

File: R/pkg/R/mllib.R (hunk 1 above)
Comment (yanboliang) [inline R/pkg/R/mllib.R]: Updated. The ```min``` and ```max``` value is correct. See the updated summary output on the above PR description.

File: R/pkg/inst/tests/testthat/test_mllib.R (hunk 4 above)
Comment (yanboliang) [inline R/pkg/inst/tests/testthat/test_mllib.R]: Here we just test R native ```summary``` works by checking the value of ```baseSummary$deviance```, we did not compared with output from other functions.

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

**Pull request closed without merging**